File: python/src/algorithms/grafs/dag.py
# Топологическая сортировка ориентированного графа —
# это такое упорядочивание всех вершин этого графа,
# что если существует ребро (v,w), то в итоговой последовательности
# v располагается раньше, чем w. Из этого определения можно вывести следствие:
# если в графе есть путь из некоторой вершины s до некоторой другой вершины t,
# то топологическая сортировка должна поставить
# s раньше, чем t.

from src.data_structures.stack import Stack

order = Stack()  # В этом стеке будет записан порядок обхода.
color = ["white", "white", "white"]  # Пример заполнения значением 'white'


def top_sort(v: int) -> None:
    color[v] = "gray"
    outgoing_edges = get_outgoing_edges(v)
    for w in outgoing_edges:
        if color[w] == "white":
            top_sort(w)
    color[v] = "black"
    order.push(v)  # Кладём обработанную вершину в стек.


def main_top_sort() -> None:
    for i in range(len(color)):
        if color[i] == "white":
            top_sort(i)


def get_outgoing_edges(v: int) -> list[int]:
    # Реализация получения исходящих рёбер для вершины v
    return []


# Пример использования
main_top_sort()

File: go/algorithms/grafs/dag.go
package graf

var order []int // В этом срезе будет записан порядок обхода.
var dag_color []string

func getOutgoingEdges(v int) []int {
	// Реализация получения исходящих рёбер для вершины v.
	return []int{}
}

func initialize(numVertices int) {
	order = make([]int, 0)
	dag_color = make([]string, numVertices)
	for i := range dag_color {
		dag_color[i] = "white"
	}
}

func topSort(v int) {
	dag_color[v] = "gray"
	outgoingEdges := getOutgoingEdges(v)
	for _, w := range outgoingEdges {
		if dag_color[w] == "white" {
			topSort(w)
		}
	}
	dag_color[v] = "black"
	order = append(order, v) // Кладём обработанную вершину в срез.
}

func mainTopSort() {
	for i := 0; i < len(dag_color); i++ {
		if dag_color[i] == "white" {
			topSort(i)
		}
	}
}

func main_dag() {
	numVertices := 10 // Задайте нужное количество вершин.
	initialize(numVertices)

	mainTopSort()
}
